openmpi: 3.1.3 -> 4.0.0

markuskowa opened the pull request:
###### Motivation for this change
This update removes some deprecated MPI-1 symbols. This breaks some applications still using these outdated symbols. Fortunately it is a simple function name change in most cases.

Upgrade guide: https://www.open-mpi.org/software/ompi/major-changes.php
Removed symbols: https://www.open-mpi.org/faq/?category=mpi-removed

CC potentially affected maintainers @costrouc @abbradar @bzizou 

###### Things done

* Add upstream patch to `mpi4py`
* patch deprecated functions in `scalapack`
* Add upstream patch to `globalarrays`
* Add upstream patches to `hfd5-mpi`
* added an entry to the release notes

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
       -> ~+3MB
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

markuskowa committed:
openmpi: 3.1.3 -> 4.0.0

openmpi-4.0.0 removes deprecated MPI-1 symbols.
For upgrade guide see:
https://www.open-mpi.org/faq/?category=mpi-removed

markuskowa committed:
globalarrays: add openmpi-4 compatibility patches

markuskowa committed:
scalapack: add openmpi4 compatibility patch

markuskowa committed:
openmpi: Use external hwloc and libevent

Recommended in the upgrade guide:
https://www.open-mpi.org/software/ompi/major-changes.php

markuskowa committed:
release-notes: add note about openmpi-4.0.0 upgrade.

Add link to openmpi upgrade guide, regarding deprecated MPI-1 symbols.

markuskowa committed:
pythonPackages.mpi4py: add patch for openmpi-4.0.0 compatibility

markuskowa commented:
@GrahamcOfBorg build python36Packages.mpi4py python36Packages.tensorflow
@GrahamcOfBorg build lammps-mpi dl-poly-classic-mpi siesta-mpi quantum-espresso-mpi

dotlambda reviewed (changes requested):
You should add the `pythonPackages.` prefix to the last commit message.

markuskowa committed:
scalapack: update meta data

* Set platform from linux -> x86_64-linux
  Scalapack does not seem to run properly on ARM (failing tests)

* Add markuskowa as maintainer

markuskowa commented:
Maybe I beat the timeout this time
@GrahamcOfBorg build scalapack

markuskowa committed:
openmpi: add patch for mpi tests on machines with less than 4 cores

Fix a bug that ignores OMPI_MCA_rmaps_base_oversubscribe (upstream patch).
This bug breaks the test from libs, such as scalapack,
on machines with less than 4 cores.

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-19-03-feature-freeze/1950/30

ryantm commented on pkgs/development/libraries/globalarrays/default.nix:
```diff
@@ -15,6 +15,8 @@ in stdenv.mkDerivation {
     sha256 = "07i2idaas7pq3in5mdqq5ndvxln5q87nyfgk3vzw85r72c4fq5jh";
   };
 
+  patches = [ ./openmpi4.patch  ];
```
Can you add a comment with a link to this patch being submitted upstream or an explanation for why this patch is specific to nixpkgs?

markuskowa replied:
I did replace this patch with upstream patches.

ryantm commented on pkgs/development/libraries/science/math/scalapack/default.nix:
```diff
@@ -12,6 +12,8 @@ stdenv.mkDerivation rec {
     sha256 = "0p1r61ss1fq0bs8ynnx7xq4wwsdvs32ljvwjnx6yxr8gd6pawx0c";
   };
 
+  patches = [ ./openmpi4.patch ];
```
Can you add a comment with a link to this patch being submitted upstream or an explanation for why this patch is specific to nixpkgs?

markuskowa replied:
scalapack is not very actively developed. I will see if can get submit this patch upstream. The patch is self written.

markuskowa replied:
Maybe to clarify on that question: according to repology there is only linuxbrew which has openmpi-4 and scalapack packaged but I do not see a solution to the problem there. We are probably the first distro to package both and have the problem solved.

markuskowa committed:
globalarrays: use autoreconfHook

markuskowa commented:
@GrahamcOfBorg build scalapack

FRidh reviewed (approved):
Looks good to me. In case we find packages that we can't trivially fix we can still include 3.1.3, but that's for then.

markuskowa committed:
hdf5: add openmpi-4 compatibility patches

markuskowa commented:
@FRidh That was my first thought to include the old version as well but I am optimistic that it should not be necessary.

markuskowa commented:
@GrahamcOfBorg build hf5-mpi

markuskowa commented:
The PR would be ready for merging.

costrouc commented:
@markuskowa thank you thank you for this work. I will greatly benefit from this PR.
